Propose Pack Destructuring & Pack Splitting (#2796)
swiftlang/swift-evolution: proposals/NNNN-pack-destructuring-and-pack-splitting.md

# SE-NNNN: Pack Destructuring & Pack Splitting

* Proposal: SE-0XXX  
* Authors: Oguz Yuksel
* Review Manager: TBD  
* Status: Draft  

> 

Reviewer: ```suggestion
* Review: ([pitch](https://forums.swift.org/t/pitch-pack-destructuring-pack-splitting/79388))

```

## Introduction  

> Swift’s variadic generics story (SE‑0393 “Parameter Packs” and SE‑0404 “Pack Iteration”) unlocks powerful abstractions over arbitrary arities. However, two missing features—recursive decomposition of packs in patterns and split‑expansion of packs at call sites—force library authors into brittle workarounds like fixed‑arity overloads or type erasure. This proposal completes the variadic‑generic toolbox by introducing:

Reviewer: ```suggestion
Swift’s variadic generics story ([SE‑0393 “Parameter Packs”](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0393-parameter-packs.md) and [SE‑0408 “Pack Iteration”](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0408-pack-iteration.md)) unlocks powerful abstractions over arbitrary arities. However, two missing features—recursive decomposition of packs in patterns and split‑expansion of packs at call sites—force library authors into brittle workarounds like fixed‑arity overloads or type erasure. This proposal completes the variadic‑generic toolbox by introducing:
```

Author: Apologies, @PsychoH13 . I unintentionally performed a force update without reviewing your suggestions. Could you please resubmit them?

// … up to Zip10
```

### Combine.zip After  
```swift
struct Zip<repeat each S>: Publisher {
  typealias Output = (repeat (each S).Output)

  let publishers: (repeat each S)

  func receive<Sub: Subscriber>(subscriber: Sub) where Sub.Input == Output {
    let (first, repeat each rest) = publishers
    first.receive(ZipHelper(subscriber, rest))
  }
}
```

**Result**: ~250 LOC → ~20 LOC, no fixed arity limit, compile‑time safety preserved.

## Future Directions  
- **Multi‑Pack Operations**  
  ```swift
  func zip(_ s1: repeat each S1, _ s2: repeat each S2) -> (repeat each (S1, S2))
  ```
- **Named Pack Elements**  
  ```swift
  let (header: repeat each headers, body: repeat each bodies) = httpResponse
  ```
- **Variadic async/await**  
  ```swift
  async let (firstResult, repeat each partials) = fetchAll(repeat each requests)
  ```

## Alternatives Considered  

### 1. Fixed‑Arity Overloads  
- Pros: trivial, no language changes  
- Cons: boilerplate, arbitrary limits, code bloat  

## References  
1. [SE‑0393: Parameter Packs](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md) – Introduces variadic type‐parameter packs.  

> 2. [SE‑0404: Pack Iteration](https://github.com/apple/swift-evolution/blob/main/proposals/0404-pack-iteration.md) – Enables iteration over packs.  

Reviewer: Pack Iteration is proposal 408 not 404. You do get a 404 when you click that link though.
```suggestion
2. [SE‑0404: Pack Iteration](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0408-pack-iteration.md) – Enables iteration over packs.  
```
